src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/DefaultHybridCache.StampedeStateT.cs

```diff
@@ -175,29 +176,64 @@ private async Task BackgroundFetchAsync()
                 // nothing from L2; invoke the underlying data store
                 if ((Key.Flags & HybridCacheEntryFlags.DisableUnderlyingData) == 0)
                 {
-                    var cacheItem = SetResult(await _underlying!(_state!, SharedToken).ConfigureAwait(false));
-
-                    // note that at this point we've already released most or all of the waiting callers; everything
-                    // else here is background
-
-                    // write to L2 if appropriate
-                    if ((Key.Flags & HybridCacheEntryFlags.DisableDistributedCacheWrite) == 0)
+                    // invoke the callback supplied by the caller
+                    T newValue = await _underlying!(_state!, SharedToken).ConfigureAwait(false);
+
+                    // If we're writing this value *anywhere*, we're going to need to serialize; this is obvious
+                    // in the case of L2, but we also need it for L1, because MemoryCache might be enforcing
+                    // SizeLimit (we can't know - it is an abstraction), and for *that* we need to know the item size.
+                    // Likewise, if we're writing to a MutableCacheItem, we'll be serializing *anyway* for the payload.
+                    //
+                    // Rephrasing that: the only scenario in which we *do not* need to serialize is if:
+                    // - it is an ImmutableCacheItem
+                    // - we're writing neither to L1 nor L2
+
+                    const HybridCacheEntryFlags DisableL1AndL2 = HybridCacheEntryFlags.DisableLocalCacheWrite | HybridCacheEntryFlags.DisableDistributedCacheWrite;
+                    var cacheItem = CacheItem;
```

Comment:
ditto

Reply:
addressed in [5e3bdba](https://github.com/dotnet/extensions/pull/5420/commits/5e3bdbafb8f2b59c811c7d66fb225a11ef379e2a)